[lld][AArch64][ELF][PAC] Support `.relr.auth.dyn` section

kovdan01 opened the pull request:
This re-applies #87635 after the issue described in https://github.com/llvm/llvm-project/pull/87635#issuecomment-2155318065 is fixed in ebc123e0793a1cbcb69b4af1548e339e018ffff2. A corresponding test is also added.

Original PR description below.

Support `R_AARCH64_AUTH_RELATIVE` relocation compression as described in https://github.com/ARM-software/abi-aa/blob/main/pauthabielf64/pauthabielf64.rst#relocation-compression.

kovdan01 committed:
[lld][AArch64][ELF][PAC] Support `.relr.auth.dyn` section

This re-applies #87635 with fix of issue described in
https://github.com/llvm/llvm-project/pull/87635#issuecomment-2155318065.

We need to discard `rel(a).dyn` section when output is a relocatable object
file. The section is always empty in such a case (as well as both auth and
regular relr sections), and emitting that resulted in its `sh_info` equal to 0.
Section with zero index is always a NULL section according to ELF spec,
and lld is unable to run `ObjFile<ELFT>::getRelocTarget` against
relocation section with `sh_info` equal to zero (the ELF spec does not
seem to define behavior for such a corner case). This is only an issue
with relocatable object file output - having `sh_info` equal to zero for
shared objects is OK.

Original commit description below.

Support `R_AARCH64_AUTH_RELATIVE` relocation compression as described in
https://github.com/ARM-software/abi-aa/blob/main/pauthabielf64/pauthabielf64.rst#relocation-compression.

kovdan01 committed:
[NFC][lld][ELF] Remove unused `sec` param of `ObjFile<ELFT>::getRelocTarget`

zeroomega commented:
I confirm it no longer generate empty `.rel.dyn` or invalid sh_info on REL section when building an relocatable object file.

zeroomega reviewed (approved):
LGTM from my point of view. I compared it with the original implementation in #87635 and confirm the empty and invalid `.rel.dyn` section in relocatable object file was fixed. 
Please wait for @MaskRay's approval before landing it.

MaskRay reviewed (changes requested):
`.rela.dyn` is currently created outside of the `config->hasDynSymTab` condition.
In relocatable links, `.rela.dyn` will be discarded by `removeUnusedSyntheticSections`.
The reverted `.relr.auth.dyn` patch incorrectly retained the empty `.relr.dyn`.

I think we should suppress the creation of `.rela.dyn` in the first place.
I will try updating `createSyntheticSections`.

The `!config->relocatable` condition in `removeUnusedSyntheticSections` in this PR can then be dropped.

MaskRay commented on lld/test/ELF/aarch64-reloc-pauth.s:
```diff
@@ -39,71 +40,171 @@
+# RUN: llvm-mc -filetype=obj -triple=aarch64 empty-rela.s -o empty-rela.o
+# RUN: ld.lld -pie -z pack-relative-relocs empty-rela.o -o empty-rela
+# RUN: llvm-readelf -S -d -r empty-rela | FileCheck --check-prefixes=EMPTY-RELA %s
+# RUN: ld.lld -r -z pack-relative-relocs empty-rela.o -o empty-rela-relocatable
```
`-o empty-rela.ro`

kovdan01 replied:
Renamed, see 87d2e3f1894a3296a04f79a69385daabbc92884e

kovdan01 committed:
Merge branch 'main' into lld-relr-fix

kovdan01 committed:
Address review comments

kovdan01 commented:
> I think we should suppress the creation of `.rela.dyn` in the first place.
> I will try updating `createSyntheticSections`.

@MaskRay I agree that it's better not to construct the section at all if we know it's not needed than construct an empty one and then delete it, thanks for implementing that it ebc123e0793a1cbcb69b4af1548e339e018ffff2.

> The `!config->relocatable` condition in `removeUnusedSyntheticSections` in this PR can then be dropped.

Deleted, thanks. I also updated the PR description.

kovdan01 committed:
Merge branch 'main' into lld-relr-fix

kovdan01 commented:
Pushed a merge commit to re-trigger buildkite job which failed to unrelated reasons
